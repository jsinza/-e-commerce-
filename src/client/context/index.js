import React, { createContext, useReducer } from 'react';
import { reducer, INITIAL_STATE } from '../reducers';
import { fetchFailure, fetchItemDetailStart, fetchItemDetailSuccess, fetchItemsStart, fetchItemsSuccess } from '../actions';
import * as itemService from '../services';

const StoreContext = createContext();

const StoreProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, INITIAL_STATE);
  const value = { state, dispatch };

  return (
    <StoreContext.Provider value={value}>
      {children}
    </StoreContext.Provider>
  );
};

const useStore = () => {
  const context = React.useContext(StoreContext);
  if (context === undefined) {
    throw new Error('useCount must be used within a CountProvider');
  }
  return context;
};

const searchItems = async (dispatch, query) => {
  dispatch(fetchItemsStart(query));
  try {
    const { data } = await itemService.searchItems(query);
    dispatch(fetchItemsSuccess(data));
  } catch ({ message }) {
    dispatch(fetchFailure(message));
  }
};

const getItem = async (dispatch, id) => {
  dispatch(fetchItemDetailStart());
  try {
    const { data } = await itemService.getItem(id);
    dispatch(fetchItemDetailSuccess(data));
  } catch ({ message }) {
    dispatch(fetchFailure(message));
  }
};

export { useStore, StoreProvider, searchItems, getItem };
