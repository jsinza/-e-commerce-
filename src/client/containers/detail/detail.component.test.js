
import React from 'react';
import { mount } from 'enzyme';
import 'babel-polyfill';
import * as router from 'react-router-dom';
import Detail from './detail.component';
import * as context from '../../context';

const ITEM = {
  'id': 'MLA907919730',
  'title': 'Cubre Zapato Zapatilla Silicona Impermeable Lluvia Acuaticas',
  'price': {
    'currency': 'ARS',
    'amount': 549,
    'decimals': 0,
  },
  'picture': 'http://http2.mlstatic.com/D_643035-MLA45049201013_032021-O.jpg',
  'condition': 'new',
  'free_shipping': false,
};

jest.mock('../../context');

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: jest.fn(),
}));

const ITEM_ID = 'SEJFEIZANME';
describe('Detail Component', () => {
  let wrapper;
  beforeEach(() => {
    router.useParams.mockReturnValue({ id: ITEM_ID });
    context.useStore.mockReturnValue({ state: { isLoading: false, selectedItem: ITEM, categories: [] } });
    wrapper = mount(<Detail />);
  });
  it('should render Detail Component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should get the search parameters', () => {
    expect(context.getItem.mock.calls[1][1]).toBe(ITEM_ID);
  });

  it('should render the home container and item title', () => {
    expect(wrapper.find("[className='detail']")).toHaveLength(1);
    expect(wrapper.find("[className='detail__description']")).toHaveLength(1);
  });

});
