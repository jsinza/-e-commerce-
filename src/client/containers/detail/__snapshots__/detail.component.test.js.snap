// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail Component should render Detail Component 1`] = `ReactWrapper {}`;
